File: javascript/basicos/validParentheses.js
// function solution(str_to_validate) {
//   const myArray = [];
//   for (let i = 0; i < str_to_validate.length; i++) {
//     if (
//       str_to_validate[i] == "(" ||
//       str_to_validate[i] == "[" ||
//       str_to_validate[i] == "{"
//     ) {
//       myArray.push(str_to_validate[i]);
//     } else if (
//       str_to_validate[i] == "]" &&
//       myArray[myArray.length - 1] == "["
//     ) {
//       myArray.pop();
//     } else if (
//       str_to_validate[i] == "}" &&
//       myArray[myArray.length - 1] == "{"
//     ) {
//       myArray.pop();
//     } else if (
//       str_to_validate[i] == ")" &&
//       myArray[myArray.length - 1] == "("
//     ) {
//       myArray.pop();
//     } else {
//       return false;
//     }
//   }
//   return true;
// }

// console.log(solution("]"));

function solutionTwo(str_to_validate) {
  stack = [];
  const mapping = {
    ")": "(",
    "}": "{",
    "]": "[",
  };

  for (let i = 0; i < str_to_validate.length; i++) {
    if (mapping.hasOwnProperty(str_to_validate[i])) {
      if (stack.length && mapping[str_to_validate[i] === stack.pop()]) {
      }
      continue;
    }
    stack.push(str_to_validate[i]);
  }
  return stack.length === 0;
}

console.log(solutionTwo("[(){}[]()]"));

File: python/basicos/valid_parentheses.py
def solution_one(str_to_validate:str):
    my_array = []
    for letter in str_to_validate:
        if letter == '(' or letter == '[' or letter == '{':
            my_array.append(letter)
        elif len(my_array) == 0:
            return False
        elif letter == ']' and my_array[-1] == '[':
            my_array.pop()
        elif letter == '}' and my_array[-1] == '{':
            my_array.pop()
        elif letter == ')' and my_array[-1] == '(':
            my_array.pop()
        else:
            return False
    return True


print(solution_one(']'))


def solution_two(str_to_validate:str):
    stack = []

    mapping = {
        ')': '(',
        '}': '{',
        ']': '['
    }

    for s in str_to_validate:
        if s in mapping:
            if len(stack) and mapping[s] == stack.pop():
                continue
        stack.append(s)
    return len(stack) == 0

print(solution_two(']'))